onnxruntime/core/framework/session_state.cc

```diff
@@ -156,20 +186,138 @@ static int64_t CalculateMemoryPatternsKey(const std::vector<std::reference_wrapp
   return key;
 }
 
-const MemoryPatternGroup* SessionState::GetMemoryPatternGroup(
-    const std::vector<std::reference_wrapper<const TensorShape>>& input_shapes) const {
+namespace {
+Status ResolveDimParams(const GraphViewer& graph, const std::map<std::string, TensorShape>& feeds, std::unordered_map<std::string, int64_t>& out) {
+  for (size_t i = 0; i < graph.GetInputs().size(); ++i) {
+    auto* input = graph.GetInputs()[i];
+    auto* shape = input->Shape();
+    auto it = feeds.find(input->Name());
+    if (it == feeds.end())
+      return Status(ONNXRUNTIME, FAIL, "Graph input " + input->Name() + " is not found in the feed list, unable to resolve the value for dynamic shape.");
+    if (!shape || shape->dim_size() != static_cast<int>(it->second.NumDimensions()))
+      return Status(ONNXRUNTIME, FAIL, "Graph input " + input->Name() +
+                                           "'s shape is not present or its shape doesn't match feed's shape."
+                                           "Unable to resolve the value for dynamic shape");
+    for (int k = 0; k < shape->dim_size(); ++k) {
+      if (shape->dim()[k].has_dim_param()) {
```

Comment:
Neither of these loops seem to need the counter so the following may be simpler:
  for (const auto*input : graph.GetInputs())
  for (const auto& dim : shape->dim())

Reply:
looks like we still need the second counter as "it" is using it

---
In reply to: [397023575](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397023575) [](ancestors = 397023575)